src/linalg/impl_linalg.rs

```diff
@@ -699,6 +701,32 @@ unsafe fn general_mat_vec_mul_impl<A, S1, S2>(
     }
 }
 
+
+/// Kronecker product of 2D matrices.
+///
+/// The kronecker product of a LxN matrix A and a MxR matrix B is a (L*M)x(N*R)
+/// matrix K formed by the block multiplication A_ij * B.
+pub fn kron<'a, A, S1, S2>(a: &ArrayBase<S1, Ix2>, b: &'a ArrayBase<S2, Ix2>) -> ArrayBase<OwnedRepr<A>, Ix2>
+where
+    S1: Data<Elem = A>,
+    S2: Data<Elem = A>,
+    A: LinalgScalar,
+    A: std::ops::Mul<&'a ArrayBase<S2, Ix2>, Output = ArrayBase<OwnedRepr<A>, Ix2>>,
+{
+    let dimar = a.shape()[0];
+    let dimac = a.shape()[1];
+    let dimbr = b.shape()[0];
+    let dimbc = b.shape()[1];
+    let mut out: Array2<MaybeUninit<A>> = Array2::uninit((dimar * dimbr, dimac * dimbc));
+    Zip::from(out.exact_chunks_mut((dimbr, dimbc)))
+        .and(a)
+        .for_each(|out, a| {
+            (*a * b).assign_to(out);
```

Comment:
We'd prefer to not create intermediate array `a * b` here. `Zip::from(a).and(&b).and(&mut out).for_each(|&a, &b, out| *out = a * b)` is an alternative?

Reply:
Hm, I don't think that quite works. I want to multiply every element of `a` by `b`, and assign it to the window of `out` that corresponds to that element of `a`.

What you wrote is element-wise multiplication, is it not?

Reply:
Also I've found difficulty getting anything but `x.assign_to(y)` to work when `y` has elements that are `MaybeUninit`, is there a better way to make the types happy around that? Maybe `AssignElem` should be implemented so that one can assign an array with elements of `T` to one of `MaybeUninit<T>`?

Reply:
> I want to multiply every element of a by the full b matrix, and assign it to the window of out that corresponds to that element of a (and is the size of the b matrix.

This should do it without the temporary, shouldn't it?

```rust
Zip::from(out.exact_chunks_mut((dimbr, dimbc)))
    .and(a)
    .for_each(|out, &a| {
        Zip::from(out).and(b).for_each(|out, &b| {
            *out = MaybeUninit::new(a * b);
        });
    });
```

> Maybe AssignElem should be implemented so that one can assign an array with elements of T to one of MaybeUninit<T>?

I think this is already the case, c.f. https://docs.rs/ndarray/0.15.3/ndarray/trait.AssignElem.html#impl-AssignElem%3CT%3E-for-%26%27a%20mut%20MaybeUninit%3CT%3E

But it isn't necessary here as we are not in generic context, i.e. we know that the left-hand side is `&mtu MaybeUninit<A>` and therefore we can just wrap the right-hand side using `MaybeUninit::new` unconditionally.

Reply:
Thank you. There are a few ways to use maybeuninit, like that, that's good, the assigbelwm trait, or creating a raw array view and working with elements as *mut A

Reply:
Good to know re using `MaybeUninit`! I was scratching my head when initially using `uninit`.

And thank you Adam for the suggestion of the nested Zips, that makes a lot of sense, I'll add that :)